librz/core/ctypes.c

```diff
@@ -1004,44 +1066,99 @@ RZ_IPI bool rz_types_open_file(RzCore *core, const char *path) {
 			char *error_msg = NULL;
 			int result = rz_type_parse_string_stateless(typedb->parser, tmp, &error_msg);
 			if (result && error_msg) {
-				eprintf("%s", error_msg);
+				RZ_LOG_ERROR("%s", error_msg);
 				free(error_msg);
 			}
 			free(tmp);
 		}
 	} else {
 		if (!rz_file_exists(path)) {
-			eprintf("File \"%s\" does not exist\n", path);
+			RZ_LOG_ERROR("File \"%s\" does not exist\n", path);
 			free(homefile);
 			return false;
 		}
 		char *error_msg = NULL;
 		int result = rz_type_parse_file_stateless(typedb->parser, path, dir, &error_msg);
 		if (result && error_msg) {
-			eprintf("%s", error_msg);
+			RZ_LOG_ERROR("%s", error_msg);
 			free(error_msg);
 		}
 	}
 	free(homefile);
 	return true;
 }
 
-RZ_IPI void rz_types_open_editor(RzCore *core, const char *typename) {
+static RzBaseType *types_get_compound_type(RzCore *core, RZ_NONNULL const char *typename) {
+	rz_return_val_if_fail(typename, NULL);
 	RzTypeDB *typedb = core->analysis->typedb;
-	char *str = rz_core_cmd_strf(core, "tc %s", typename ? typename : "");
-	char *tmp = rz_core_editor(core, "*.h", str);
-	if (tmp) {
-		char *error_msg = NULL;
-		int result = rz_type_parse_string_stateless(typedb->parser, tmp, &error_msg);
-		if (result) {
-			// TODO: remove previous types and save new edited types
-			//rz_type_db_purge(typedb);
-		}
+	RzBaseType *t = rz_type_db_get_base_type(typedb, typename);
+	if (!t) {
+		RZ_LOG_ERROR("Cannot find type \"%s\"\n", typename);
+		return false;
+	}
+	if (t->kind == RZ_BASE_TYPE_KIND_ATOMIC) {
+		RZ_LOG_ERROR("Cannot edit atomic type \"%s\"\n", typename);
+		return false;
+	}
+	return t;
+}
+
+static bool types_change_type_str(RzCore *core, RzBaseType *t, RZ_NONNULL const char *old, RZ_NONNULL const char *new) {
+	rz_return_val_if_fail(t && old && new, false);
+	RzTypeDB *typedb = core->analysis->typedb;
+	// Remove the original type first
+	rz_type_db_delete_base_type(typedb, t);
+	char *error_msg = NULL;
+	int result = rz_type_parse_string_stateless(typedb->parser, new, &error_msg);
+	if (result) {
 		if (error_msg) {
-			eprintf("%s\n", error_msg);
+			RZ_LOG_ERROR("%s\n", error_msg);
+		}
+		free(error_msg);
+		// There is an error during the parsing thus we restore the old type
+		int result = rz_type_parse_string_stateless(typedb->parser, old, &error_msg);
```

Comment:
What about not using `rz_type_db_delete_base_type()` to delete the base type entirely, but just removing it from the hashtable without freeing to re-insert it later?
Using the `old` string means that it must be guaranteed that printing a type to C and then parsing it again will always result in exactly the original type, which I would not trust if not strictly necessary.